compiler/rustc_data_structures/src/vec_cache.rs

```diff
@@ -0,0 +1,349 @@
+
+struct Slot<V> {
+    // We never construct &Slot<V> so it's fine for this to not be in an UnsafeCell.
+    value: V,
+    // This is both an index and a once-lock.
+    //
+    // 0: not yet initialized.
+    // 1: lock held, initializing.
+    // 2..u32::MAX - 2: initialized.
+    index_and_lock: AtomicU32,
+}
+
+/// This uniquely identifies a single `Slot<V>` entry in the buckets map, and provides accessors for
+/// either getting the value or putting a value.
+#[derive(Copy, Clone, Debug)]
+struct SlotIndex<V> {
+    // the index of the bucket in VecCache (0 to 20)
+    bucket_idx: usize,
+    // number of entries in that bucket
+    entries: usize,
+    // the index of the slot within the bucket
+    index_in_bucket: usize,
+
+    bucket_ty: PhantomData<V>,
+}
+
+impl<V> SlotIndex<V> {
+    // This makes sure the counts are consistent with what we allocate, precomputing each bucket a
+    // compile-time. Visiting all powers of two is enough to hit all the buckets.
+    //
+    // We confirm counts are accurate in the slot_index_exhaustive test.
+    const ENTRIES_BY_BUCKET: [usize; 21] = {
+        let mut entries = [0; 21];
+        let mut key = 0;
+        loop {
+            let si = SlotIndex::<V>::from_index(key);
+            entries[si.bucket_idx] = si.entries;
+            if key == 0 {
+                key = 1;
+            } else if key == (1 << 31) {
+                break;
+            } else {
+                key <<= 1;
+            }
+        }
+        entries
+    };
+
+    // This unpacks a flat u32 index into identifying which bucket it belongs to and the offset
+    // within that bucket. As noted in the VecCache docs, buckets double in size with each index.
+    // Typically that would mean 31 buckets (2^0 + 2^1 ... + 2^31 = u32::MAX - 1), but to reduce
+    // the size of the VecCache struct and avoid uselessly small allocations, we instead have the
+    // first bucket have 2**12 entries. To simplify the math, the second bucket also 2**12 entries,
+    // and buckets double from there.
+    //
+    // We assert that [0, 2**32 - 1] uniquely map through this function to individual, consecutive
+    // slots (see `slot_index_exhaustive` in tests). Note that in practice this mapping truncates
+    // upper indices on 32-bit systems, since each entry is at least 4 bytes and so we don't have
+    // enough address space to store the whole range.
+    #[inline]
+    const fn from_index(idx: u32) -> Self {
+        let mut bucket = match idx.checked_ilog2() {
+            Some(x) => x as usize,
+            None => 0,
+        };
+        let entries;
+        let running_sum;
+        if bucket <= 11 {
+            entries = 1 << 12;
+            running_sum = 0;
+            bucket = 0;
+        } else {
+            entries = 1 << bucket;
+            running_sum = entries;
+            bucket = bucket - 11;
+        }
+        let max = match (isize::MAX as usize).checked_div(std::mem::size_of::<Slot<V>>()) {
```

Comment:
when can this `checked_div`  ever return `None`? it should neither overflow or divide by 0